File: solutions/kotlin/luhn/1/src/main/kotlin/Luhn.kt
object Luhn {

    fun isValid(candidate: String): Boolean {
       val cand = candidate.replace(" ","")
       if (cand.length <= 1)
           return false
       var sum = 0
       for(i in cand.length-1 downTo 0){
           val char = cand[i]
           if (char.isDigit()){
               var d = char.digitToInt()
               if(((cand.length-i) % 2 ==0) && d!=9) 
                       d = 2*d % 9
               sum += d    
           } else return false
       }
       return sum % 10 == 0
    }
}

File: solutions/dart/luhn/1/lib/luhn.dart
import 'package:collection/collection.dart';

class Luhn {
  bool valid(String candidate){
    var cand = candidate.replaceAll(' ','');
    var nonDigits = new RegExp(r'[^0-9]');
    if(!nonDigits.allMatches(cand).isEmpty)
      return false;
    if(cand.length<=1)
      return false;
    var checksum = cand.split('')
      .map((d)=>int.parse(d))
      .mapIndexed((idx,d) 
                   => (d!=9 && (cand.length - idx) % 2 == 0)
                        ?  (2*d)% 9 : d)
      .toList()
      .reduce((a,b) => a + b);
    return checksum % 10 == 0;
  }
}
